src/UCTNode.cpp

```diff
@@ -150,41 +137,34 @@ void UCTNode::link_nodelist(std::atomic<int> & nodecount,
     }
 
     // sort (this will reverse scores, but linking is backwards too)
-    std::sort(begin(nodelist), end(nodelist));
-
-    // link the nodes together
-    auto childrenadded = 0;
+    std::stable_sort(rbegin(nodelist), rend(nodelist));
 
     LOCK(get_mutex(), lock);
 
     for (const auto& node : nodelist) {
-        auto vtx = new UCTNode(node.second, node.first, init_eval);
-        link_child(vtx);
-        childrenadded++;
+        m_children.emplace_back(std::make_unique<UCTNode>(node.second, node.first, init_eval));
     }
 
-    nodecount += childrenadded;
+    nodecount += m_children.size();
     m_has_children = true;
 }
 
 void UCTNode::kill_superkos(KoState & state) {
-    UCTNode * child = m_firstchild;
-
-    while (child != nullptr) {
+    for (auto& child : m_children) {
         int move = child->get_move();
-
         if (move != FastBoard::PASS) {
             KoState mystate = state;
             mystate.play_move(move);
 
             if (mystate.superko()) {
-                UCTNode * tmp = child->m_nextsibling;
-                delete_child(child);
-                child = tmp;
+                // TODO delete comment before merge.
+                // Instead of removing invalid node just mark node invalid.
+                // This logic is already used in UCTSearch.cpp, I want to
+                // revisit ko moves next so it can be cleaned up more then.
+                child->invalidate();
```

Comment:
Could not removing a child make problems if anyone depends on a check with `has_children()`?

Reply:
has_children instead looks at m_has_children (see L~315) which measures if the expand happened.
plus there will ALWAYS be at least one valid child (the pass child), except in get_no_pass_child (which returns a potentially null pointer to UCTNode